### PR TITLE
Adds saved query endpoints.

### PR DESCRIPTION
**Pull request by andela-Kshittu**

### CODE REVIEW

**Comment by andela-Kshittu** [comment]
Do not merge until https://github.com/datadotworld/platform/pull/5138 has been approved.

File: src/main/resources/world/data/api/swagger.json [hunk 1]
```diff
@@ -49,6 +49,32 @@
...
+        "content": {
+          "description": "Query body.",
+          "type": "string"
+        },
+        "type": {
+          "description": "Type of language in which this query is written. Can be either 'SPARQL' or 'SQL'.",
+          "type": "string"
+        },
+        "published": {
+          "description": "Is query published.",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: There should be more text here to describe what `published` actually means.

File: src/main/resources/world/data/api/swagger.json [hunk 2]
```diff
@@ -49,6 +49,32 @@
       "title": "Project Create Response",
       "type": "object"
     },
+    "CreateQueryRequest": {
+      "type": "object",
+      "properties": {
+        "name": {
+          "description": "Query name",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Missing period at end of text.

File: src/main/resources/world/data/api/swagger.json [hunk 3]
```diff
@@ -49,6 +49,32 @@
...
+          "type": "string"
+        },
+        "content": {
+          "description": "Query body.",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: `Query body` isn't as clear as it should be. Explain that this is where the actual query to be executed in the specified language is set.

File: src/main/resources/world/data/api/swagger.json [hunk 4]
```diff
@@ -1474,12 +1500,33 @@
           "description": "Parameter type. Can be either 'literal' or 'iri'",
           "type": "string"
         },
-        "dataType": {
+        "datatype": {
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: revert this change.

File: src/main/resources/world/data/api/swagger.json [hunk 5]
```diff
@@ -1912,6 +1959,14 @@
       "required": true,
       "type": "string"
     },
+    "queryid": {
+      "description": "Query unique identifier for saved queries.",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Where is there a `queryid` and `queryId`? In the public API we don't follow the private API parameter style of using lowercase `id`s.

File: src/main/resources/world/data/api/swagger.json [hunk 6]
```diff
@@ -2886,18 +2941,30 @@
...
-    "/datasets/{owner}/{id}/sync": {
-      "get": {
-        "description": "Same as `POST:/dataset/{owner}/{id}/sync`.",
-        "operationId": "syncViaGet",
+      },
+      "post": {
+        "consumes": [
+          "application/json"
+        ],
+        "description": "Create saved queries associated with a dataset.",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Create a saved query in a specified dataset.

File: src/main/resources/world/data/api/swagger.json [hunk 7]
```diff
@@ -2907,10 +2974,19 @@
           "200": {
             "description": "**OK**\nThe request has succeeded.",
             "examples": {
-              "application/json": "{\n  \"message\": \"Sync started.\"\n}"
+              "application/json": {
+                "body": "DECLARE metric AS BOOLEAN = true\nSELECT *\nFROM table\nLIMIT 5000",
+                "created": "2018-03-30T02:30:54.751Z",
+                "id": "d529a67d-a27a-4f88-86f0-c8c3003afed9",
+                "language": "SQL",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: why is it called `language` here and `type` in the request?

File: src/main/resources/world/data/api/swagger.json [hunk 8]
```diff
@@ -2957,20 +3027,25 @@
             ]
           }
         ],
-        "summary": "Sync files (via GET)",
+        "summary": "Create a saved query for a dataset",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Create a saved query in a specified dataset.

File: src/main/resources/world/data/api/swagger.json [hunk 9]
```diff
@@ -2957,20 +3027,25 @@
...
         ]
-      },
-      "post": {
-        "description": "Sync files within a dataset. This method will process the latest data available for files added from URLs or via streams.",
-        "operationId": "sync",
+      }
+    },
+    "/datasets/{owner}/{id}/queries/{queryid}": {
+      "delete": {
+        "description": "Delete a saved query for a dataset",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Delete a saved query in a specified dataset.

File: src/main/resources/world/data/api/swagger.json [hunk 10]
```diff
@@ -3031,16 +3099,17 @@
             ]
           }
         ],
-        "summary": "Sync files",
+        "summary": "Delete a saved query",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Delete a saved query in a specified dataset.

File: src/main/resources/world/data/api/swagger.json [hunk 11]
```diff
@@ -3031,16 +3099,17 @@
...
-    "/datasets/{owner}/{id}/v/{versionId}": {
-      "get": {
-        "description": "Retrieve a version of a dataset. The definition of the dataset will be returned, not its data.",
-        "operationId": "getDatasetByVersion",
+      },
+      "put": {
+        "consumes": [
+          "application/json"
+        ],
+        "description": "Update a saved query for a dataset.",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Update a saved query in a specified dataset.

File: src/main/resources/world/data/api/swagger.json [hunk 12]
```diff
@@ -3162,32 +3188,26 @@
         "security": [
           {
             "oauth": [
-              "user_api_read"
+              "user_api_write"
             ]
           }
         ],
-        "summary": "Retrieve a dataset version",
+        "summary": "Update a saved query for a dataset",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Update a saved query in a specified dataset.

File: src/main/resources/world/data/api/swagger.json [hunk 13]
```diff
@@ -4789,6 +5099,261 @@
         "tags": [
           "queries"
         ]
+      },
+      "post": {
+        "consumes": [
+          "application/json"
+        ],
+        "description": "Create a saved query for a project.",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Create a saved query in a specified project.

File: src/main/resources/world/data/api/swagger.json [hunk 14]
```diff
@@ -4789,6 +5099,261 @@
...
+          }
+        },
+        "security": [
+          {
+            "oauth": [
+              "user_api_write"
+            ]
+          }
+        ],
+        "summary": "Create a saved query for a project",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Create a saved query in a specified project.

File: src/main/resources/world/data/api/swagger.json [hunk 15]
```diff
@@ -4789,6 +5099,261 @@
...
+        "tags": [
+          "queries"
+        ]
+      }
+    },
+    "/projects/{owner}/{id}/queries/{queryid}": {
+      "delete": {
+        "description": "Delete a saved query for a project",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Delete a saved query in a specified project.

File: src/main/resources/world/data/api/swagger.json [hunk 16]
```diff
@@ -4789,6 +5099,261 @@
...
+          }
+        },
+        "security": [
+          {
+            "oauth": [
+              "user_api_write"
+            ]
+          }
+        ],
+        "summary": "Delete a saved query for a project",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Delete a saved query in a specified project.

File: src/main/resources/world/data/api/swagger.json [hunk 17]
```diff
@@ -4789,6 +5099,261 @@
...
+        "tags": [
+          "queries"
+        ]
+      },
+      "put": {
+        "consumes": [
+          "application/json"
+        ],
+        "description": "Update a saved query for a project.",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Update a saved query in a specified project.

File: src/main/resources/world/data/api/swagger.json [hunk 18]
```diff
@@ -4789,6 +5099,261 @@
...
+          }
+        },
+        "security": [
+          {
+            "oauth": [
+              "user_api_write"
+            ]
+          }
+        ],
+        "summary": "Update a saved query for a project",
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Update a saved query in a specified project.

**Comment by andela-Kshittu** [comment]
@nirajpatel I see your concern, the git diff view is a bit confusing. Only query endpoints were touched, I did a diff check with IntelliJ to double check that.

File: src/main/resources/world/data/api/swagger.json (hunk 4 above)
Comment (andela-Kshittu) [reply src/main/resources/world/data/api/swagger.json]: That's what we have here https://github.com/datadotworld/platform/blob/123e6e6e06d40986a3fc17ee673da8da99b872ff/domain/src/main/java/world/data/api/dtos/pub/QueryParameter.java#L17

Leaving it as `dataType` breaks the `SwaggerConsumerTest`

File: src/main/resources/world/data/api/swagger.json (hunk 7 above)
Comment (andela-Kshittu) [reply src/main/resources/world/data/api/swagger.json]: Cos `language` is what we have in QuerySummaryResponse. The field `type` in QueryFetchDto is mapped to field language `language` https://github.com/datadotworld/platform/blob/678d983bbd84a532c4ac4f30747edf64d2c8139f/dw-api/public-api/src/main/java/world/data/api/pub/mappers/QueryMapper.java#L29 . 

Note : the field has now been renamed to `language` in the request objects too.

File: src/main/resources/world/data/api/swagger.json [hunk 19]
```diff
@@ -1474,12 +1500,33 @@
...
+          "description": "Query name",
+          "type": "string"
+        },
+        "content": {
+          "description": "Query body.",
+          "type": "string"
+        },
+        "published": {
+          "description": "Is query published.",
+          "type": "boolean"
```
Comment (nirajpatel) [inline src/main/resources/world/data/api/swagger.json]: Update these descriptions according to what you have for `CreateQueryRequest`

Comment (andela-Kshittu) [reply src/main/resources/world/data/api/swagger.json]: Ah...I missed that. Thanks.
